Fix an error when using `RSpec/FilePath` and revert to enabled by default

ydah opened the pull request:
Fix: https://github.com/rubocop/rubocop-rspec/issues/1717

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [ ] Added tests.
- [x] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

- [ ] Added the new cop to `config/default.yml`.
- [ ] The cop is configured as `Enabled: pending` in `config/default.yml`.
- [ ] The cop is configured as `Enabled: true` in `.rubocop.yml`.
- [ ] The cop documents examples of good and bad code.
- [ ] The tests assert both that bad code is reported and that good code is not reported.
- [ ] Set `VersionAdded: "<<next>>"` in `default/config.yml`.

If you have modified an existing cop's configuration options:

- [x] Set `VersionChanged: "<<next>>"` in `config/default.yml`.

ydah commented:
It appears that Danger is failing because GITHUB_TOKEN does not have write permission.
I don't have permission to see the Settings in this repository, so could someone please check if it is set as follows?

https://github.com/rubocop/rubocop-rspec/settings/actions
![スクリーンショット 2023-09-12 18 30 32](https://github.com/rubocop/rubocop-rspec/assets/13041216/919acfc1-a094-45d1-87dc-e88c7f18dfe5)

pirj commented on config/obsoletion.yml:
```diff
@@ -27,9 +27,3 @@ renamed:
   RSpec/FactoryBot/FactoryNameStyle: FactoryBot/FactoryNameStyle
   RSpec/FactoryBot/SyntaxMethods: FactoryBot/SyntaxMethods
-
-split:
```
Maybe keep this commented out?
Is there a better way to remember to do it on a major release?

bquorning replied:
We can set up a milestone in GitHub and add an issue to it.

bquorning replied:
I added [the milestone](https://github.com/rubocop/rubocop-rspec/milestone/2), but didn’t open a issue.

pirj commented on docs/modules/ROOT/pages/cops_rspec.adoc:
```diff
@@ -2012,6 +2012,12 @@ expect { my_app.print_report }.to output('Hello World').to_stdout
 
+This cop is deprecated.
+We plan to remove it in the next major version update to 3.0.
+The migration destinations are `RSpec/SpecFilePathSuffix`
```
Targets?

pirj reviewed (approved):
Thank you!

bquorning commented:
> It appears that Danger is failing because GITHUB_TOKEN does not have write permission.

On the surface, that seems a bit dangerous. Isn’t is possible to use a token with fewer permissions.

pirj commented:
Is Danger trying to tell us that we’re introducing a cop in Enabled status?

wondering if there’s an option to tell danger not to post PR comments but rather just dump their warnings to stdout for inspection

splattael commented on CHANGELOG.md:
```diff
@@ -2,9 +2,11 @@
 ## 2.24.0 (2023-09-08)
 
-- Split `RSpec/FilePath` into `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`. `RSpec/FilePath` cop is enabled by default, the two new cops are pending and need to be enabled explicitly. ([@ydah])
+- **(Breaking)** Split `RSpec/FilePath` into `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`. `RSpec/FilePath` cop is disabled by default and the two new cops are pending and need to be enabled explicitly. ([@ydah])
```
I am confused. Isn't `RSpec/FilePath` now enabled by default again? :thinking: If so, is this really a breaking change then? :nerd_face: 

```suggestion
- Split `RSpec/FilePath` into `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`. `RSpec/FilePath` cop is enabled by default and the two new cops are pending and need to be enabled explicitly. ([@ydah])
```

ydah replied:
Oh, sure, we'll enable it back, but isn't it is breaking change for those who use v2.24.0?

bquorning replied:
Perhaps we’re asking “does two breaking changes equal zero breaking changes”?

ydah replied:
Ah, I understand. Indeed it is.

ydah replied:
I updated this PR. How about it?

domon-envato replied:
Feedback: Our builds were actually broken twice by this two breaking changes.

ydah commented:
> Is Danger trying to tell us that we’re introducing a cop in Enabled status?
> 
> wondering if there’s an option to tell danger not to post PR comments but rather just dump their warnings to stdout for inspection

https://danger.systems/ruby/
No functionality seems to be provided for output to stdout.

ydah commented:
@rubocop/rubocop-rspec Danger replacement should be able to be done as a follow-up, so it would be nice to be able to merge and do a patch version release if there are no problems with this first. How about it?

pirj reviewed (approved):
Yhank you!

bquorning commented on CHANGELOG.md:
```diff
@@ -2,9 +2,11 @@
 
 ## Master (Unreleased)
 
+- Fix an error when using `RSpec/FilePath` and revert to enabled by default. If you have already migrated `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`, no problem. The `RSpec/FilePath` before migration and the `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat` at the destination are available respectively. ([@ydah])
```
Isn’t it a problem for people had custom configuration for `RSpec/FilePath`, and changed it when upgrading to v2.24.0?

ydah replied:
Indeed it is. Explicitly disabled `RSpec/FilePath`. How about this?

pirj commented on CHANGELOG.md:
```diff
@@ -2,9 +2,11 @@
 
 ## Master (Unreleased)
 
+- Fix an error when using `RSpec/FilePath`. If you have already migrated `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`, no problem. The `RSpec/FilePath` before migration and the `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat` at the destination are available respectively. If you use `RSpec/FilePath` as is, add `RSpec/FilePath: Enabled: true` to your .rubocop.yml. ([@ydah])
```
Destination -> target?

ydah replied:
😱 Oh, my apologize.

pirj reviewed (commented):
Now when I think of this - why have we disabled the cop? Should we enable it again for those who don’t follow the brave ‘NewCops: true’ path? Can we split and remove the old cop in a major version, and meanwhile just keep those two new pending cops?
New cops would duplicate the function of the old one, but this is not a problem, is it?

ydah commented:
I had misunderstood the comment I received at https://github.com/rubocop/rubocop-rspec/pull/1718#discussion_r1327976868 . I changed it to `Enabled: false` and changed the migration plan for CHANGELOG. How about this?

pirj commented:
Just for the reference https://github.com/rubocop/rubocop-rspec/issues/1717#issuecomment-1715201355

i don’t think we need to disable FilePath until v3 (where we remove it).

ydah commented on config/default.yml:
```diff
@@ -448,7 +448,7 @@ RSpec/ExpectOutput:
 RSpec/FilePath:
   Description: Checks that spec file paths are consistent and well-formed.
-  Enabled: false
+  Enabled: true
```
I updated this.

pirj commented on CHANGELOG.md:
```diff
@@ -2,9 +2,11 @@
 
 ## Master (Unreleased)
 
+- Fix an error when using `RSpec/FilePath` and revert to enabled by default. If you have already moved to `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`, disable `RSpec/FilePath` explicitly as `Enabled: false`. The `RSpec/FilePath` before migration and the `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat` at the target are available respectively. ([@ydah])
```
```suggestion
- Fix an error when using `RSpec/FilePath` and revert to enabled by default. If you have already moved to `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat`, disable `RSpec/FilePath` explicitly as `Enabled: false`. The `RSpec/FilePath` before migration and the `RSpec/SpecFilePathSuffix` and `RSpec/SpecFilePathFormat` as the target are available respectively. ([@ydah])
```

ydah replied:
I updated this PR.

pirj reviewed (approved):
Thank you!

timon commented:
I think I see regression with this and custom inflections in Zeitwerk.
rubocop-rspec 2.24.1 fails CI for our project using customer inflections:
```ruby
# config/initializers/zeitwerk.rb
Rails.autoloaders.each do |autoloader|
  autoloader.inflector = Zeitwerk::Inflector.new
  autoloader.inflector.inflect(
    'bvd' => 'BvD'
  )
end
```

Every spec file that covers `BvD` module now registers an offense like this:
```
spec/services/bvd/adapters/fetch_data_spec.rb:5:1: C: RSpec/FilePath: Spec path should end with bv_d/adapters/fetch_data*_spec.rb.
describe BvD::Adapters::FetchData do
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It was not an issue neither for `2.24.0`, nor for `2.23.2` (although I'm not sure if it was skipped)

AlexWayfer commented:
@timon I see no info that `rubocop-rspec` can support Zeitwerk inflections for it's own configuration.

I can suggest to you to add some configuration.

You've got the warning from `RSpec/FilePath`. Let's see its documentation: https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#rspecfilepath

`CustomTransform` option looks exactly what do you need.

domon-envato commented on lib/rubocop/cop/rspec/file_path.rb:
```diff
@@ -5,6 +5,12 @@ module Cop
+      # We plan to remove it in the next major version update to 3.0.
+      # The migration targets are `RSpec/SpecFilePathSuffix`
+      # and `RSpec/SpecFilePathFormat`.
+      # If you are using this cop, please plan for migration.
```
@ydah What is the suggested way to migrate?

I've updated my config like the following for `0.24.0`.

* Removed `RSpec/FilePath`
* Disabled `RSpec/SpecFilePathFormat`
* Enabled `RSpec/SpecFilePathSuffix`

But this new change in `0.24.1` is forcing me to update my config again like the following.

* Add `RSpec/FilePath` back and disable it

Does that mean my config will be broken when `RSpec/FilePath` is removed in the next major version update to 3.0?

pirj replied:
I understand your frustration with some reconfiguration required for a minor/patch level version dependency updates.

May I kindly ask you to elaborate what you mean by "forcing"?

You meant 2.24.0.

> my config will be broken ... in the next major version update to 3.0?

Major updates often come with breaking changes, so most probably yes.
But worry not, we'll provide a detailed update checklist like we did with [1.x -> 2.x update](https://docs.rubocop.org/rubocop-rspec/upgrade_to_version_2.html).

domon-envato replied:
Thanks for responding to my question 🙏 

> May I kindly ask you to elaborate what you mean by "forcing"?
> You meant 2.24.0.

* Before `2.24.0`: I had `RSpec/FilePath` explicitly disabled in my config.
* At `2.24.0`: I had to remove `RSpec/FilePath` completely from my config or the rubocop command would not exit with a successful status.
* At `2.24.1`: Now I had to disable `RSpec/FilePath` again or it would be enabled by default.

After updated the config twice in each of these two versions, I was a bit alerted by the words **"plan for migration"** and curious about what that actually means.

But on second thought, I think `2.24.1` just moved the state back to what we were at before `2.24.0`.

If I interpret what you said correctly, there will be no breaking changes like this until 3.0. But we probably will need to update the config when upgrading to 3.x, which is quite normal. Is my understanding correct?

pirj replied:
I see, thanks for the explanation. And please accept my apologies for introducing breaking changes.

Your understanding is correct.
